### PR TITLE
CORE-5091 Consensual Transaction signing improvements

### PR DESCRIPTION
**Pull request by vlajos**
* CORE-7031 `SignableData` to sign metadata with txId and usage of it.
* CORE-7031 Sign with all available missing keys in the Receive Finality Flow.
* CORE-7095 Use PlatformInfo service.
* CORE-7095 Prepare Metadata in TxBuilder Factory.
* Support multiple CPIs, static member lists in smoke tests and use them for a full consensual ledger flow test.
* Rename consensual flow test cordapp
* Some `Map` -> `LinkedHashMap` conversion to make serialization deterministic.

Its API part: https://github.com/corda/corda-api/pull/647

### CODE REVIEW

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt [hunk 1]
```diff
@@ -75,12 +81,19 @@ class ConsensualFinalityFlow(
             log.debug { "Received signature from ${session.counterparty} for signed transaction ${signedTransaction.id}" }
 
             try {
+                if(signature.by != sessionPublicKeys[idx]) {
+                    throw CordaRuntimeException(
+                        "A session with ${session.counterparty} sent back a signature with another keys." +
```
Comment (lankydan) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: "sent back a signature with another keys." -> "sent back a signature with another key."

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt [hunk 2]
```diff
@@ -59,13 +65,13 @@ class ConsensualFinalityFlow(
 
         var signedByParticipantsTransaction = signedTransaction
 
-        for (session in sessions) {
+        sessions.forEachIndexed{idx, session ->
```
Comment (lankydan) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: Formatting.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualReceiveFinalityFlow.kt [hunk 3]
```diff
@@ -44,16 +44,10 @@ class ConsensualReceiveFinalityFlow(
...
-            signingService.sign(
-                bytes = signedTransaction.id.bytes,
-                publicKey = memberLookup.myInfo().ledgerKeys.first(),
-                signatureSpec = SignatureSpec.ECDSA_SHA256
-            )
-        )
+        val txWithMySignature = signedTransaction.sign(memberLookup.myInfo().ledgerKeys.first())
+        val mySignature = txWithMySignature.signatures.last()
+
+        session.send(mySignature)
```
Comment (lankydan) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualReceiveFinalityFlow.kt]: Would prefer:
```
val transactionWithSignature = signedTransaction.sign(memberLookup.myInfo().ledgerKeys.first())
session.send(transactionWithSignature.signatures.last())
```

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualReceiveFinalityFlow.kt [hunk 4]
```diff
@@ -63,15 +57,17 @@ class ConsensualReceiveFinalityFlow(
                     "${signedTransaction.id} instead"
         }
 
+        //TODO this looks like a candidate for signedTx.verify() or verifySignatures at least.
         for (signature in signedTransactionToFinalize.signatures) {
             try {
                 // TODO Signature spec to be determined internally by crypto code
                 // Throws if the signature is invalid which will terminate the peer flows with an error
+                val signedData = SignableData(transactionId, signature.metadata)
```
Comment (lankydan) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualReceiveFinalityFlow.kt]: I think we can construct this here instead:

```
clearData = serializationService.serialize(SignableData(transactionId, signature.metadata)).bytes
```

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualReceiveFinalityFlow.kt [hunk 5]
```diff
@@ -63,15 +57,17 @@ class ConsensualReceiveFinalityFlow(
...
                 digitalSignatureVerificationService.verify(
                     publicKey = signature.by,
                     signatureSpec = SignatureSpec.ECDSA_SHA256,
                     signatureData = signature.signature.bytes,
-                    clearData = signedTransactionToFinalize.id.bytes
+                    clearData = serializationService.serialize(signedData).bytes
```
Comment (lankydan) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualReceiveFinalityFlow.kt]: We used to have `HashingService` in DP1 that would serialize and sign an object. It sounds like we may re-add this in the near future.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt [hunk 6]
```diff
@@ -75,12 +81,19 @@ class ConsensualFinalityFlow(
...
 
             try {
+                if(signature.by != sessionPublicKeys[idx]) {
+                    throw CordaRuntimeException(
+                        "A session with ${session.counterparty} sent back a signature with another keys." +
+                                "Expected: ${sessionPublicKeys[idx]} vs Received: ${signature.by}"
+                    )
+                }
                 // TODO Do not hardcode signature spec
+                val signedData = SignableData(signedTransaction.id, signature.metadata)
```
Comment (lankydan) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: I think we can construct this here instead:

```
clearData = serializationService.serialize(SignableData(transactionId, signature.metadata)).bytes
```

File: libs/ledger/ledger-common-internal/src/main/kotlin/net/corda/ledger/common/internal/transaction/TransactionSignatureUtils.kt [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import net.corda.v5.application.crypto.DigitalSignatureMetadata
+import net.corda.v5.application.crypto.SigningService
+import net.corda.v5.application.serialization.SerializationService
+import net.corda.v5.base.annotations.Suspendable
+import net.corda.v5.crypto.SecureHash
+import net.corda.v5.crypto.SignatureSpec
+import java.security.PublicKey
+import java.time.Instant
+
+//TODO(Should this be another service? Or could we just squeeze this into the signingservice?)
```
Comment (lankydan) [inline libs/ledger/ledger-common-internal/src/main/kotlin/net/corda/ledger/common/internal/transaction/TransactionSignatureUtils.kt]: Definitely not the signing service. I'd make a `TransctionSigningService` if needed.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualReceiveFinalityFlow.kt (hunk 3 above)
Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualReceiveFinalityFlow.kt]: This section changed a bit due to the refactor of signedTransaction.sign(...)

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualReceiveFinalityFlow.kt (hunk 4 above)
Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualReceiveFinalityFlow.kt]: I thought it is a bit more self-documenting/readable in two steps.
But happy to change if you prefer more concise.

File: libs/ledger/ledger-common-internal/src/main/kotlin/net/corda/ledger/common/internal/transaction/TransactionSignatureUtils.kt (hunk 7 above)
Comment (vlajos) [reply libs/ledger/ledger-common-internal/src/main/kotlin/net/corda/ledger/common/internal/transaction/TransactionSignatureUtils.kt]: Maybe we can come back to this after the crypto API's changes. CORE-6969

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ConfigTests.kt [hunk 8]
```diff
@@ -9,6 +9,7 @@ import org.junit.jupiter.api.Order
 import org.junit.jupiter.api.Test
 
 @Order(40)
+@Suppress("FunctionName")
```
Comment (lankydan) [inline applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ConfigTests.kt:12]: Why is this needed now if nothing changed?

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt [hunk 9]
```diff
@@ -48,14 +58,44 @@ class ConsensualSignedTransactionImpl(
...
         ConsensualLedgerTransactionImpl(this.wireTransaction, serializer)
 
-    /** CORE-5091 it does not do anything at the moment. */
-    override fun addSignature(publicKey: PublicKey): ConsensualSignedTransaction {
-        return ConsensualSignedTransactionImpl(serializer, wireTransaction, signatures)
+    @Suspendable
+    override fun addSignature(publicKey: PublicKey): Pair<ConsensualSignedTransaction, DigitalSignatureAndMetadata> {
+        val newSignature = createTransactionSignature(signingService, serializer, getCpiSummary(), id, publicKey)
+        return Pair(ConsensualSignedTransactionImpl(
+            serializer, signingService, digitalSignatureVerificationService, wireTransaction,
```
Comment (lankydan) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt]: formatting

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt [hunk 10]
```diff
@@ -48,14 +58,44 @@ class ConsensualSignedTransactionImpl(
...
+            signatures + newSignature
+        ), newSignature)
     }
 
+    override fun addSignature(signature: DigitalSignatureAndMetadata): ConsensualSignedTransaction =
```
Comment (lankydan) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt:70]: What is the difference between `addSignature` here and `ConsensualTransactionBuilder.sign`? Should they both be called `sign`?

**Review by lankydan — COMMENTED** [review]
Small things.

File: applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ConfigTests.kt (hunk 8 above)
Comment (vlajos) [reply applications/workers/workers-smoketest/src/smokeTest/kotlin/net/corda/applications/workers/smoketest/ConfigTests.kt:12]: Not exactly needed, just suppresses the intellij formatting grey underlines caused by the test's names.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt (hunk 10 above)
Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt:70]: Functionality-wise, there is not difference. @mnesbit mentioned prefered to use `addSignature`. The reasoning was that `sing()` could be confusing for some reason. (what I forgot...)

Comment (lankydan) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt:70]: Mmmk, personally i think they should be the same since they are both signing things when you pass in your public key.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt (hunk 9 above)
Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt]: I am not too sure if I get exactly what you meant, but added some more line breaks.

Comment (lankydan) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt]: Yeah, the formatter doesn't always deal with this. But, when things don't fit onto a single line, having multiple linebreaks instead is correct.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt [hunk 11]
```diff
@@ -74,31 +82,40 @@ class ConsensualFinalityFlow(
...
-                digitalSignatureVerificationService.verify(
-                    publicKey = signature.by,
-                    signatureSpec = SignatureSpec.ECDSA_SHA256,
-                    signatureData = signature.bytes,
-                    clearData = signedTransaction.id.bytes
+            val expectedSigningKeys = signatures.map { it.by }
+            if (expectedSigningKeys.toSet() != sessionPublicKeys[idx].toSet()) {
+                throw CordaRuntimeException(
+                    "A session with ${session.counterparty} did not back the expected keys" +
+                            "Expected: ${sessionPublicKeys[idx]} vs Received: $expectedSigningKeys"
```
Comment (blsemo) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: That is the wrong way round (expected vs actual).
Also the whole message is a bit hard to understand, are you missing the main verb (i.e. give back or bring back)?

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt [hunk 12]
```diff
@@ -46,16 +56,52 @@ class ConsensualSignedTransactionImpl(
...
+        ConsensualSignedTransactionImpl(serializationService, signingService, digitalSignatureVerificationService,
+            wireTransaction, signatures + signature)
+
     override fun getMissingSigningKeys(): Set<PublicKey> {
         val alreadySigned = signatures.map{it.by}.toSet()
         val requiredSigningKeys = this.toLedgerTransaction().requiredSigningKeys
         return requiredSigningKeys.filter { !it.isFulfilledBy(alreadySigned) }.toSet()
     }
+
+    override fun verifySignatureValidity() {
```
Comment (blsemo) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt]: I don't like the name of that method. It verifies that we have all the required signatures and they are all valid, whereas the name suggest it validates a single signature.

How about `verifySignatures()`?

**Review by blsemo — COMMENTED** [review]
Looks mostly good,, a couple of comments.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt (hunk 11 above)
Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: Thanks.  I have changed the wording and fixed the order.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt (hunk 12 above)
Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt]: I have changed it to that.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt (hunk 11 above)
Comment (lankydan) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: I don't think "vs" is a good choice of words either. We should use "but was" or "but received" instead imo.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt (hunk 12 above)
Comment (lankydan) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/transaction/ConsensualSignedTransactionImpl.kt]: I agree, was trying to find what interface provided this method earlier on today but got sidetracked.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt (hunk 11 above)
Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: Thanks. Updated.

Comment (lankydan) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: Reading the code, I don't the message is the wrong way around. It is because `expectedSigningKeys` is named wrong. It should be `signingKeys` or `receivedSigningKeys`, these are from the signatures received from the peers. `sessionPublicKeys` are defined at the start based on the passed in sessions. So surely `sessionPublicKeys` are the "expected keys" and what is currently `expectedSigningKeys" are actually the "received signing keys".

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt [hunk 13]
```diff
@@ -59,13 +67,13 @@ class ConsensualFinalityFlow(
 
         var signedByParticipantsTransaction = signedTransaction
 
-        for (session in sessions) {
+        sessions.forEachIndexed { idx, session ->
```
Comment (lankydan) [inline libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: Upon thinking about this more, I know its unlikely to cause any issues, but I'd be more comfortable if `sessionPublicKeys` is a `Map<Session, List<PublicKey>>` or `Map<MemberX500Name, List<PublicKey>>` rather than using the index in the returned list. Especially if we move to using the batch send/receives, this logic is more likely to fail or require rework anyway.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt (hunk 11 above)
Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: Hmm. You are right. I have updated it again.

File: libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt (hunk 13 above)
Comment (vlajos) [reply libs/ledger/ledger-consensual-impl/src/main/kotlin/net/corda/ledger/consensual/impl/flows/finality/ConsensualFinalityFlow.kt]: I have updated it.

**Review by lankydan — DISMISSED** [review]

**Review by lankydan — APPROVED** [review]

**Review by ronanbrowne — APPROVED** [review]
LGTM (Build)
